### ISSUE TITLE
Syntax sugar

### ISSUE THREAD

**Issue by kg** [WebAssembly/design#386]
We've been introducing a bunch of s-expression syntax sugar lately, where certain constructs get mapped by the parser/lexer into other constructs. This breaks the property where s-expressions are a mostly direct representation of the AST and you can round-trip them (previously 100% possible, even with binary encoding).

Are we generally in agreement that these syntax sugar mechanisms are intended to let us iterate and experiment or do we want to keep them long-term?

I can see the syntax sugar being really valuable for keeping compilers simple, but it destroys one of the early principles we had where the s-expressions are a simple 'view source' format that you can round-trip, with unsophisticated parsing and stringifying rules. We're throwing that out with our current approach.

In particular, having a lot of syntax sugar implies that most (or all) debugging tools will end up having to implement decompilers that reconstruct the sugar (the way many .net and java debugging tools tend to work serves as evidence of this). Are we okay with paying these costs because we think it's worth it? I'm pretty uncomfortable with parsing & encoding becoming this complicated.

**Comment by rossberg-chromium** [comment]
Just to clarify: this is completely orthogonal to tooling. Compilers and
tools are free not to desugar, and often shouldn't. It's purely a
spec-internal device, and a standard approach in factoring formal language
specs. It simplifies specifying and reasoning about a language, by reducing
the number of cases you have to consider.

**Comment by kg** [comment]
'free not to desugar' means we've forked the syntax, though, and that the test suite is no longer usable without relying on the set of extra desugaring magic (essentially, a compiler) that ml-proto uses. I think that is a sensible decision to make, but I wouldn't be happy with it.

**Comment by lukewagner** [comment]
We've all agreed (and [documented](https://github.com/WebAssembly/design/blob/master/TextFormat.md#official-text-format) that the s-expr language is temporary so I'm confused why we're discussing long-term aspects of s-expressions like tooling and View Source.

When it comes to designing the real text format, though, I do expect we'll want some amount of sugar to enhance reading and writing since that is the primary use case for having a text format.

**Comment by kg** [comment]
This is not purely a text format consideration (and I still don't buy the claim that we'll magically get rid of sexprs at some point - this implies rewriting the test suite...), because we're introducing a huge disconnect between *whatever* text format we have and the actual opcodes and binary format. Previously, the text format was assumed to be a near-1:1 representation of the binary format, which made it a suitable place to do all our prototyping and iteration. We're turning it into a (simple, to be fair) programming language with a compiler, containing lots of things that no longer correspond to the binary encoding or to the AST that decoders will be dealing with. That makes me nervous.

**Comment by lukewagner** [comment]
> (and I still don't buy the claim that we'll magically get rid of sexprs at some point - this implies
> rewriting the test suite...)

Or automatically translate.  If we specify the text format with a function `to_text : Ast.module_ -> string`, then this should be especially easy.

**Comment by JSStats** [comment]
I appears that recognising some operations a 'sugar' is a recognition that: 1) they are not really needed, 2) that producers are likely to emit different patterns for the same effective operation.

The 'sugar' seems like a macro layer to me. It might be useful for some producers, particularly hand written code, but consumers that actually need to match any of these pattern will need to match both the 'sugar' of 'desugar' versions (or canonicalize the code by expanding the macros) which just adds a burden for them. If it is a 'web' requirement to be able to view the source then the binary format should include these macros as well as user defined macros, and perhaps this is what is being achieved here.

**Comment by rossberg-chromium** [comment]
There still seems to be some misunderstanding. I'm sorry if I haven't made
this sufficiently clear. Let me try again.

What we call "desugaring" here is an implementation detail of the spec. The
difference between core syntax and enriched syntax is _not_ externally
observable. Both the text format (official or inofficial) and the binary
format will of course provide the same complete "sugared" set of opcodes.

I'm currently working on a patch to move the actual desugaring out of the
parser to the spec directory, to make clearer what the complete set of
spec'ed opcodes is.

@kg, does that address your concerns?

**Comment by kg** [comment]
Yes. If the sugar is still defined as opcodes, that lets you do the 1-1 mapping everywhere other than the VM (if you want) which satisfies my concerns - the sugar becomes an implementation detail for the opcodes and not something the whole ecosystem is concerned with. I guess I was confused about this since so much of it happens in the parsing/lexing/etc stage right now.

In practice they do become a narrow set of macros, though, so it may make sense to phase them out eventually in favor of macros that compilers can use if they need them - but for now it certainly makes sense to just do the sugared opcodes so that we aren't waiting for macros to be figured out.

**Comment by AndrewScheidecker** [comment]
> I still don't buy the claim that we'll magically get rid of sexprs at some point

There's an amusing story that might be relevant: [McCarthy on the origin of S-expressions](http://www-formal.stanford.edu/jmc/history/lisp/node3.html)

> Writing eval required inventing a notation representing LISP functions as LISP data, and such a notation was devised for the purposes of the paper with no thought that it would be used to express LISP programs in practice.
> ...
> S.R. Russell noticed that eval could serve as an interpreter for LISP, promptly hand coded it, and we now had a programming language with an interpreter.
> ...
> The unexpected appearance of an interpreter tended to freeze the form of the language, and some of the decisions made rather lightheartedly for the ``Recursive functions ...'' paper later proved unfortunate.
> ...
>Another reason for the initial acceptance of awkwardnesses in the internal form of LISP is that we still expected to switch to writing programs as M-expressions. The project of defining M-expressions precisely and compiling them or at least translating them into S-expressions was neither finalized nor explicitly abandoned. It just receded into the indefinite future, and a new generation of programmers appeared who preferred internal notation to any FORTRAN-like or ALGOL-like notation that could be devised.

**Comment by lukewagner** [comment]
One thing working in our favor is that ultimately it is only the binary format which must be natively decoded by the browsers and there's not a chance the s-expr will accidentally become the binary format.  The browser only cares about the text format for the non-semantic job of generating text for devtools, View Source, etc.  Even if we develop a first cohort of tools that all produce/consume s-exprs, and thus run the risk of a de facto standard emerging, if engines start generating the "real" text format (which they can do at any time, and with good competitive reason if it provides the nicer dev experience), other tools would follow (at least the common ones we expect people to use, like the assembler).

**Comment by titzer** [comment]
Having automatic tools to translate between forms also helps. If we decide
that we hate all existing s-expressions, then we can do
s-expressions->binary->newformat in one fell swoop.

**Comment by titzer** [comment]
The s-expression format should be seen as temporary for now, so let's deal with syntactic sugar issues when we're in binary format territory.
